### PR TITLE
DOMEvent support for CustomEvent

### PR DESCRIPTION
**Pull request by keul**
See #3708

### CODE REVIEW

**Comment by philippjfr** [comment]
This seems completely reasonable to me. My only concern here is that `detail` may contain certain objects which can't be naively serialized to JSON and will therefore error, e.g. `Undefined` objects are one example that I've frequently seen issues with.

File: panel/models/event-to-object.ts [hunk 1]
```diff
@@ -25,6 +25,11 @@ SOFTWARE.
 export function serializeEvent(event: any): any {
   const data: any = {};
 
+  // support for CustomEvents: the whole `detail` object is serialized
+  if (Object.getPrototypeOf(event).initCustomEvent !== undefined) {
```
Comment (mattpap) [inline panel/models/event-to-object.ts:29]: Any reason to use this specific detection code, instead of `event instanceof CustomEvent`?

Comment (keul) [reply panel/models/event-to-object.ts:29]: @mattpap yep, more simple, but it will be also true for `Event` instance.

**Comment by keul** [comment]
> This seems completely reasonable to me. My only concern here is that `detail` may contain certain objects which can't be naively serialized to JSON and will therefore error, e.g. `Undefined` objects are one example that I've frequently seen issues with.

@philippjfr uops, you are right about this.
Any suggestion on how to improve? JSON.stringify/JSON.parse would fix this but maybe is not that efficient.

File: panel/models/event-to-object.ts (hunk 1 above)
Comment (keul) [reply panel/models/event-to-object.ts:29]: … I mean: there's nothing wrong on supporting also `Event`, but it's the general design of this module seems to want to limit properties read from an event and I didn't wanted to go against this

File: panel/models/event-to-object.ts [hunk 2]
```diff
@@ -25,6 +25,11 @@ SOFTWARE.
...
+    Object.assign(data, { detail: event.detail } );
```
Comment (philippjfr) [inline panel/models/event-to-object.ts:30]: Maybe simply add a check if `detail` is undefined to the conditional.

Comment (keul) [reply panel/models/event-to-object.ts:30]: @philippjfr to resume: do you want to remove the whole poor man typechecking and just test for a `detail` attribute?

**Comment by keul** [comment]
> > This seems completely reasonable to me. My only concern here is that `detail` may contain certain objects which can't be naively serialized to JSON and will therefore error, e.g. `Undefined` objects are one example that I've frequently seen issues with.
> 
> @philippjfr uops, you are right about this. Any suggestion on how to improve? JSON.stringify/JSON.parse would fix this but maybe is not that efficient.

@philippjfr I would go for trying to serialize to JSON and abort with a console error in case it fails

File: panel/models/event-to-object.ts (hunk 2 above)
Comment (philippjfr) [reply panel/models/event-to-object.ts:30]: Probably simpler in the end tbh.

Comment (keul) [reply panel/models/event-to-object.ts:30]: Done!

Comment (philippjfr) [reply panel/models/event-to-object.ts:30]: Okay, and this does what you need?

Comment (keul) [reply panel/models/event-to-object.ts:30]: Yes, with https://github.com/holoviz/panel/pull/3709/commits/2ec39f04d857c98ca4a3dc26d4837dbf6c787325 I'm still able pass through the server serialized data using `.detail` as defined by MDN.

There's still an open issue about what to do if `detail` contains stuff that can't be converted to JSON.

**Comment by philippjfr** [comment]
> @philippjfr I would go for trying to serialize to JSON and abort with a console error in case it fails

This seems reasonable and shouldn't ever be too expensive.

**Comment by keul** [comment]
> > @philippjfr I would go for trying to serialize to JSON and abort with a console error in case it fails
> 
> This seems reasonable and shouldn't ever be too expensive.

Hi @philippjfr

I investigated the issue a little more and I misunderstood something. There's no issue on serializing an object using `JSON.stringify`. For example:

```javascript
JSON.stringify({aaa: 555, ggg: () => {}, bbb: undefined});
'{"aaa":555}'
```

But if I use this same value for `.detail`, a JavaScript error is raised: `[object Function] is not serializable`.
This came from bokeh itself, specifically calling a [to_serializable method](https://github.com/bokeh/bokeh/blob/9c677362d832dafb5c10036fd6fe53ffbb4c258c/bokehjs/src/lib/core/serializer.ts#L89).

Also: not that this serializer module seems a lot different in the [bokeh 3 branch](https://github.com/bokeh/bokeh/blob/fca16442ae90afcd2ac61f4e554e538776730831/bokehjs/src/lib/core/serialization/serializer.ts)

So: the only way to test `detail` for being valid is to import this serializer module from bokeh, but to be honest it seems too much to me.

Currently we already have a JavaScript exception with an error message. I would leave this check to bokeh internal.

**Comment by philippjfr** [comment]
For now let's just roundtrip it through `JSON.stringify` and `JSON.parse` so it drops the unserializable parts.

**Comment by philippjfr** [comment]
Thanks for your patience with this one @keul!

**Pull request merged**